keras_cv/layers/regularization/dropblock_2d.py

```diff
@@ -0,0 +1,191 @@
+            raise ValueError(
+                f"dropblock_size must be greater than 0. Received: {dropblock_size}"
+            )
+        if not 0.0 <= keep_probability <= 1.0:
+            raise ValueError(
+                f"keep_probability must be a number between 0 and 1. "
+                f"Received: {keep_probability}"
+            )
+
+        self._keep_probability = keep_probability
+        self._dropblock_size = dropblock_size
+        self._data_format = conv_utils.normalize_data_format(data_format)
+
+    def call(self, x, training=False):
+        if not training or self._keep_probability == 1.0:
+            return x
+
+        if self._data_format == "channels_last":
+            _, height, width, _ = x.get_shape().as_list()
+        else:
+            _, _, height, width = x.get_shape().as_list()
+
+        total_size = width * height
+        dropblock_size = min(self._dropblock_size, width, height)
+
+        # Seed_drop_rate is the gamma parameter of DropBlock.
+        seed_drop_rate = (
+            (1.0 - self._keep_probability)
+            * total_size
+            / dropblock_size**2
+            / ((width - self._dropblock_size + 1) * (height - self._dropblock_size + 1))
+        )
+
+        # Forces the block to be inside the feature map.
+        w_i, h_i = tf.meshgrid(tf.range(width), tf.range(height))
+        valid_block = tf.logical_and(
+            tf.logical_and(
+                w_i >= int(dropblock_size // 2), w_i < width - (dropblock_size - 1) // 2
+            ),
+            tf.logical_and(
+                h_i >= int(dropblock_size // 2), h_i < width - (dropblock_size - 1) // 2
+            ),
+        )
+
+        if self._data_format == "channels_last":
+            valid_block = tf.reshape(valid_block, [1, height, width, 1])
+        else:
+            valid_block = tf.reshape(valid_block, [1, 1, height, width])
+
+        randnoise = self._random_generator.random_uniform(tf.shape(x), dtype=tf.float32)
+        valid_block = tf.cast(valid_block, dtype=tf.float32)
+        seed_keep_rate = tf.cast(1 - seed_drop_rate, dtype=tf.float32)
+        block_pattern = (1 - valid_block + seed_keep_rate + randnoise) >= 1
+        block_pattern = tf.cast(block_pattern, dtype=tf.float32)
+
+        if self._data_format == "channels_last":
+            ksize = [1, self._dropblock_size, self._dropblock_size, 1]
+        else:
+            ksize = [1, 1, self._dropblock_size, self._dropblock_size]
+
+        block_pattern = -tf.nn.max_pool(
+            -block_pattern,
+            ksize=ksize,
+            strides=[1, 1, 1, 1],
+            padding="SAME",
+            data_format="NHWC" if self._data_format == "channels_last" else "NCHW",
+        )
+
+        percent_ones = tf.cast(tf.reduce_sum(block_pattern), tf.float32) / tf.cast(
+            tf.size(block_pattern), tf.float32
+        )
+
+        return x / tf.cast(percent_ones, x.dtype) * tf.cast(block_pattern, x.dtype)
```

Comment:
So this layer also raises activations by some percentage based on the number of ones?  interesting.  Maybe a quick comment explaining this logic in the docstring would be good.

Reply:
The normalization is a part of the algorithm in the paper, but I'm not sure what's the motivation behind it.

Reply:
I'm mentioning the normalization in the docstring. The paper nowhere explains why the values are normalized, and I cannot find any information behind the motivation.

Reply:
The paper mentions multiple times a parallel to dropout:
> DropBlock resembles dropout [ 1] when block_size = 1 and resembles SpatialDropout [20] when block_size covers the full feature map. (pg. 3, paragraph 2)

So, like in [dropout](https://github.com/tensorflow/tensorflow/blob/2faee75c9385e371baf0dd9b93c46bb20ecba8b9/tensorflow/python/ops/nn_ops.py#L5679), I assume this was done to maintain the variance of the input distribution. Otherwise we would see a change in the signal's magnitude during test time.

Reply:
@lucasdavid good find. I will add note in the comment.

Reply:
Thanks!  Let me know when you add the comment @sebastian-sz

Reply:
@LukeWood It is added on line 228:

https://github.com/sebastian-sz/keras-cv/blob/327e127402ec6bbba62421888494ac51fe09601a/keras_cv/layers/regularization/dropblock_2d.py#L228